File: packages/core/src/security/smart-filter.ts
// ============================================================
// SmartFilter — smart file filtering
// Compatible with .gitignore + built-in rules + secret file blocking
// Platform independent
// ============================================================

import * as fs from 'fs';
import * as path from 'path';

export interface FilterResult {
  allowed: string[];
  blocked: Array<{ file: string; reason: string }>;
}

const BUILTIN_IGNORES = [
  'node_modules/', '.git/', '.vscode/', '.idea/',
  'dist/', 'build/', 'out/', 'target/',
  'cache/', 'logs/', 'temp/', 'tmp/',
  '*.log', '*.tmp', '.DS_Store', 'Thumbs.db',
];

const SECRET_FILES = [
  '.env', '.env.local', '.env.production',
  '*.pem', '*.key', '*.p12', '*.pfx',
  '*.keystore', '*.jks', 'credentials.json',
  'service-account.json', 'secrets.yml', 'secrets.yaml',
];

export class SmartFilter {
  private ignorePatterns: string[];
  private blockSecrets: boolean;

  constructor(workspaceRoot: string, options?: { extraIgnores?: string[]; blockSecrets?: boolean }) {
    this.blockSecrets = options?.blockSecrets ?? true;

    // Load .gitignore
    this.ignorePatterns = [...BUILTIN_IGNORES];
    const gitignorePath = path.join(workspaceRoot, '.gitignore');
    if (fs.existsSync(gitignorePath)) {
      const content = fs.readFileSync(gitignorePath, 'utf-8');
      this.ignorePatterns.push(
        ...content.split('\n')
          .map((l) => l.trim())
          .filter((l) => l && !l.startsWith('#')),
      );
    }
    if (options?.extraIgnores) {
      this.ignorePatterns.push(...options.extraIgnores);
    }
    if (this.blockSecrets) {
      this.ignorePatterns.push(...SECRET_FILES);
    }
  }

  /** Filter a list of files */
  filter(files: string[]): FilterResult {
    const allowed: string[] = [];
    const blocked: Array<{ file: string; reason: string }> = [];

    for (const file of files) {
      if (this.shouldIgnore(file)) {
        blocked.push({ file, reason: this.getReason(file) });
      } else {
        allowed.push(file);
      }
    }

    return { allowed, blocked };
  }

  /** Decide about a single file */
  shouldIgnore(filePath: string): boolean {
    const basename = path.basename(filePath);

    for (const pattern of this.ignorePatterns) {
      if (this.matchSimple(pattern, basename) || this.matchSimple(pattern, filePath)) {
        return true;
      }
    }
    return false;
  }

  private matchSimple(pattern: string, target: string): boolean {
    // Simple glob matching (no micromatch dependency)
    if (pattern === target) return true;
    if (pattern.endsWith('/') && target.startsWith(pattern)) return true;
    if (pattern.startsWith('*.')) {
      return target.endsWith(pattern.substring(1));
    }
    if (pattern.includes('*')) {
      const regex = new RegExp(
        '^' + pattern.replace(/\./g, '\\.').replace(/\*/g, '.*') + '$',
      );
      return regex.test(target);
    }
    return target.includes(pattern);
  }

  private getReason(filePath: string): string {
    const basename = path.basename(filePath);
    if (basename.startsWith('.env')) return 'Environment variable file';
    if (basename.endsWith('.pem') || basename.endsWith('.key')) return 'Key file';
    if (filePath.includes('node_modules/')) return 'Dependency directory';
    if (filePath.includes('.git/')) return 'Git internal file';
    if (filePath.includes('dist/') || filePath.includes('build/')) return 'Build artifact';
    return 'Matched an ignore rule';
  }
}

File: packages/jetbrains-plugin/src/main/kotlin/com/gitpilot/core/SmartFilter.kt
// ============================================================
// GitPilot JetBrains — SmartFilter (Kotlin)
// Smart file filtering: .gitignore + built-in rules + secret detection
// Fully aligned with VS Code smart-filter.ts
// ============================================================

package com.gitpilot.core

import java.io.File

class SmartFilter(workspaceRoot: String, blockSecrets: Boolean = true) {

    private val ignorePatterns = mutableListOf<String>()

    init {
        // Built-in ignores
        ignorePatterns.addAll(BUILTIN_IGNORES)
        // Load .gitignore
        val gitignore = File(workspaceRoot, ".gitignore")
        if (gitignore.exists()) {
            gitignore.readLines().forEach { line ->
                val trimmed = line.trim()
                if (trimmed.isNotEmpty() && !trimmed.startsWith("#")) {
                    ignorePatterns.add(trimmed)
                }
            }
        }
        // Secret files
        if (blockSecrets) ignorePatterns.addAll(SECRET_FILES)
    }

    fun filter(files: List<String>): FilterResult {
        val allowed = mutableListOf<String>()
        val blocked = mutableListOf<Pair<String, String>>()
        for (file in files) {
            if (shouldIgnore(file)) {
                blocked.add(Pair(file, getReason(file)))
            } else {
                allowed.add(file)
            }
        }
        return FilterResult(allowed, blocked)
    }

    fun shouldIgnore(filePath: String): Boolean {
        val basename = File(filePath).name
        for (pattern in ignorePatterns) {
            if (matchSimple(pattern, basename) || matchSimple(pattern, filePath)) return true
        }
        return false
    }

    private fun matchSimple(pattern: String, target: String): Boolean {
        if (pattern == target) return true
        if (pattern.endsWith("/") && target.startsWith(pattern)) return true
        if (pattern.startsWith("*.")) return target.endsWith(pattern.substring(1))
        if (pattern.contains("*")) {
            val regex = Regex("^" + Regex.escape(pattern).replace("\\*", ".*") + "$")
            return regex.matches(target)
        }
        return target.contains(pattern)
    }

    private fun getReason(filePath: String): String {
        val basename = File(filePath).name
        return when {
            basename.startsWith(".env") -> "Environment variable file"
            basename.endsWith(".pem") || basename.endsWith(".key") -> "Key file"
            filePath.contains("node_modules/") -> "Dependency directory"
            filePath.contains(".git/") -> "Git internal file"
            filePath.contains("dist/") || filePath.contains("build/") -> "Build artifact"
            else -> "Matched an ignore rule"
        }
    }

    companion object {
        val BUILTIN_IGNORES = listOf(
            "node_modules/", ".git/", ".vscode/", ".idea/",
            "dist/", "build/", "out/", "target/",
            "cache/", "logs/", "temp/", "tmp/",
            "*.log", "*.tmp", ".DS_Store", "Thumbs.db"
        )
        val SECRET_FILES = listOf(
            ".env", ".env.local", ".env.production",
            "*.pem", "*.key", "*.p12", "*.pfx",
            "*.keystore", "*.jks", "credentials.json",
            "service-account.json", "secrets.yml", "secrets.yaml"
        )
    }
}
